narwhal/primary/src/synchronizer.rs

```diff
@@ -826,3 +863,161 @@ impl Synchronizer {
         Ok(self.inner.certificate_store.contains(&digest)?)
     }
 }
+
+/// Holds information for a suspended certificate. The certificate can be accepted into the DAG
+/// once `missing_parents` become empty.
+struct SuspendedCertificate {
+    certificate: Certificate,
+    missing_parents: HashSet<CertificateDigest>,
+    notify: broadcast::Sender<()>,
+}
+
+/// Keeps track of suspended certificates and their missing parents.
+/// The digest keys in `suspended` and `missing` can overlap, but a digest can exist in one map
+/// but not the other.
+///
+/// They can be combined into a single map, but it seems more complex to differentiate between
+/// suspended certificates that is not a missing parent of another, from a missing parent without
+/// the actual certificate.
+///
+/// Traversal of certificates that can be accepted should start from the missing map, i.e.
+/// 1. If a certificate exists in `missing`, remove its entry.
+/// 2. Find children of the certificate, update their missing parents.
+/// 3. If a child certificate no longer has missing parent, traverse from it with step 1.
+#[derive(Default)]
+struct State {
+    // Maps digests of suspended certificates to details including the certificate itself.
+    suspended: HashMap<CertificateDigest, SuspendedCertificate>,
+    // Maps digests of certificates that are not yet in the DAG, to digests of certificates that
+    // include them as parents. Keys are prefixed by round number to allow GC.
+    missing: BTreeMap<(Round, CertificateDigest), HashSet<CertificateDigest>>,
+}
+
+impl State {
+    /// Checks if a digest is suspended. If it is, gets a notification for when it is accepted.
+    fn check_suspended(&self, digest: &CertificateDigest) -> Option<AcceptNotification> {
+        self.suspended
+            .get(digest)
+            .map(|suspended_cert| new_accept_notification(suspended_cert.notify.subscribe()))
+    }
+
+    /// Inserts a certificate with its missing parents into the suspended state.
```

Comment:
Probably let's add some more context here. Which are the cases that this can fail.